Add Secure Enclave Support to iOS/macOS

ahmeddhus opened the pull request:
### Description
This PR adds support for storing keychain items using the Secure Enclave on iOS and macOS.
A new `useSecureEnclave` boolean option is introduced in `IOSOptions` and `AppleOptions`.
When enabled, secrets are stored using Secure Enclave-backed keys if available.
If Secure Enclave is unavailable or unwrapping fails, the plugin falls back to standard keychain storage.

### Changes
- Added `useSecureEnclave` to Dart options (`IOSOptions`, `AppleOptions`)
- Passed the flag to native code via method channel
- Updated iOS/macOS native implementations to use Secure Enclave when requested
- Implemented fallback logic for devices without Secure Enclave
- Updated integration tests to cover Secure Enclave and fallback scenarios
- Updated documentation and code comments


### Motivation
Addresses [#917](https://github.com/juliansteenbakker/flutter_secure_storage/issues/917)
Provides stronger security for sensitive data on supported Apple devices.


### Checklist
- [x]  All tests pass
- [x]  Documentation updated
- [x]  Backwards compatible

Commit:
Add `useSecureEnclave` boolean to ios_options.dart, apple_options.dart, and macos_options.dart

Commit:
Implement secure enclave in iOS native code for iOS and macOS

Commit:
Add unit tests for `useSecureEnclave`

Commit:
Add documentation about `useSecureEnclave` in README.md

Commit:
Add `useSecureEnclave` scenarios to app_test.dart when secure enclave is not available

Commit:
Add `useSecureEnclave` to 'IOSOptions with custom values' test in flutter_secure_storage_test.dart

Commit:
Add `useSecureEnclave` scenarios to app_test.dart when secure enclave is  available

Commit:
replace ` IOSOptions.defaultOptions` with `const IOSOptions()`

SarcasticNoodle commented on flutter_secure_storage/example/integration_test/app_test.dart:
```diff
@@ -75,6 +78,130 @@ void main() {
+      if (!(Platform.isIOS &&
+          Platform.environment.containsKey('SIMULATOR_DEVICE_NAME'))) {
+        return; // Skip when not running on iOS Simulator
+      }
```
let me know what you think about this
```suggestion
    testWidgets('Enclave requested on iOS Simulator falls back gracefully',
        skip: !(Platform.isIOS && Platform.environment.containsKey('SIMULATOR_DEVICE_NAME')),
        (WidgetTester tester) async {
```

Commit:
Use `skip` (testWidgets parameter) instead of condition inside `WidgetTesterCallback`

juliansteenbakker commented:
Thank you for your contribution! I will test this as soon as i get my apple device up and running.

juliansteenbakker committed:
Merge branch 'develop' into develop-1

juliansteenbakker commented on flutter_secure_storage/example/integration_test/app_test.dart:
```diff
@@ -78,6 +81,127 @@ void main() {
+    testWidgets(
+        'iOS device: baseline (useSecureEnclave=false) write/read/delete',
+        (WidgetTester tester) async {
+      if (!(Platform.isIOS &&
```
Why not use the `skip` parameter like the test above?

ahmeddhus replied:
Fixed

juliansteenbakker commented on flutter_secure_storage/example/integration_test/app_test.dart:
```diff
@@ -78,6 +81,127 @@ void main() {
+    testWidgets(
+        'iOS device: useSecureEnclave=true with non-prompting access control (applicationPassword) write/read/delete',
+        (WidgetTester tester) async {
+      if (!(Platform.isIOS &&
```
Why not use the `skip` parameter like the test above?

ahmeddhus replied:
Fixed

juliansteenbakker replied:
Thank you, i will merge as soon as lint issues are fixed and ci passes

ahmeddhus replied:
All green now ✅

juliansteenbakker reviewed (changes requested):
Overall implementation looks good, just a minor issue

Commit:
Use `skip` (testWidgets parameter) instead of condition inside `WidgetTesterCallback`

Commit:
Merge remote-tracking branch 'origin/develop' into develop

Commit:
fix lint issue in app_test.dart

zmuranaka commented:
@juliansteenbakker Are there any changes this needs left? My company is hoping this comes in a release soon 💪

juliansteenbakker commented:
I've created some code enabling a smooth migration from and to the secure enclave. I will add this asap, and release it as a whole later this week.

juliansteenbakker committed:
Merge branch 'develop' into develop

juliansteenbakker commented:
I have one more question right now about the implementation. It seems that on default settings, 'userPresence' is selected for AccessControlFlags. This causes a touchId prompt for every action. I got 7 prompts when opening the example app with 3 keys.

I think we should use [LAContext](https://developer.apple.com/documentation/localauthentication/lacontext) to prevent this. What do you think?

Commit:
Use shared LAContext to reuse biometrics for Secure Enclave ops

ahmeddhus commented:
Hi, I accidentally closed my PR while discarding changes in GitHub’s conflict editor.
The branch still exists, but I don’t have permission to reopen the PR.
Can you please reopen it for me?

@juliansteenbakker

juliansteenbakker commented:
HI, it says that your develop branch is up to date (the same) as my develop branch, so i can't reopen

Commit:
Merge remote-tracking branch 'origin/develop' into develop-use-LAContext

# Conflicts:
#	flutter_secure_storage_darwin/darwin/flutter_secure_storage_darwin/Sources/flutter_secure_storage_darwin/FlutterSecureStorage.swift

Commit:
Reuse performDelete for Secure Enclave key cleanup and surface errors

ahmeddhus commented:
Can you check now please?

ahmeddhus commented:
> I have one more question right now about the implementation. It seems that on default settings, 'userPresence' is selected for AccessControlFlags. This causes a touchId prompt for every action. I got 7 prompts when opening the example app with 3 keys.
> 
> I think we should use [LAContext](https://developer.apple.com/documentation/localauthentication/lacontext) to prevent this. What do you think?

Good catch. I’ve updated the iOS/macOS plugin to reuse a single LAContext per Flutter method call and pass it via kSecUseAuthenticationContext (with a 30s touchIDAuthenticationAllowableReuseDuration). This keeps the userPresence default for Secure Enclave but collapses the multiple Touch ID/Face ID prompts down to one per operation. Let me know if you want a different reuse duration or to change the default flags

juliansteenbakker commented:
Thank you for the changes! I tried the code, it now gives me 1 popup per action, however, readAll always returns zero items.

Commit:
Fix readAll to properly handle Secure Enclave items

Commit:
Add fallback if Secure Enclave read failed (no wrapped key), try plain text

ahmeddhus commented:
> Thank you for the changes! I tried the code, it now gives me 1 popup per action, however, readAll always returns zero items.

Thanks for the feedback! I’ve just pushed a fix, could you check it again and let me know if it works now?

westracer commented:
I tried to use this new feature. Writing the data works (`dataStatus` = 0), but I'm getting an error when trying to read that new entry (`SecKeyCreateDecryptedData` fails):

```
▿ Optional<CFErrorRef>
  - some : Error Domain=com.apple.LocalAuthentication Code=-1009 "ACL operation is not allowed: 'ock'" UserInfo={NSDebugDescription=ACL operation is not allowed: 'ock', NSLocalizedDescription=Operation is not allowed.}
```

Dart code:
```
secretStorage: const FlutterSecureStorage(
  aOptions: AndroidOptions(enforceBiometrics: true),
  iOptions: IOSOptions(
    accessibility: KeychainAccessibility.passcode,
    accessControlFlags: [AccessControlFlag.userPresence],
    useSecureEnclave: true,
  ),
),
...
await secretStorage.read(key: key);
...
await secretStorage.write(
  key: key,
  value: secret,
);

```

The fallback to "standard read" fails as well, obviously, since the stored data is encrypted (it returns nil string).

Tested on iPhone SE (2nd gen), iOS 18.4, with Touch ID.

westracer commented:
> I tried to use this new feature. Writing the data works (`dataStatus` = 0), but I'm getting an error when trying to read that new entry (`SecKeyCreateDecryptedData` fails):

After some research, I think the secure enclave private key needs `.privateKeyUsage` access control flag, so it can be used for crypto operations. However, passing it from the Dart side to the whole storage breaks it too 
because it creates another issue with the wrapped key (SecItemAdd returns -25293 authorization error).

Another problem is that the `userPresence` parameter is also applied to the SE private key, and it triggers user authentication 2 times on read: for the SE private key, and for the actual encrypted item (and possibly for the wrapped key as well?). I think it should be triggered only one time for the actual item, and not for the SE key.

My suggested changes are:

1. Make the SE private key use some hardcoded access control params including .privateKeyUsage (or maybe make them customizable?)
2. "deleteAll" method should delete the SE private key as well (or create another method to delete it?)

See [my commit](https://github.com/QuickBirdEng/flutter_secure_storage/commit/3d465546e1dc8c0288ba2a5c410e717a939b1a0b) for my fix implementation

<details>
<summary>Updated ensureEnclavePrivateKey</summary>

```
    private func ensureEnclavePrivateKey(service: String?) throws -> SecKey {
        let tag = enclaveKeyTag(for: service) as CFData

        let query: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrKeyType: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrApplicationTag: tag,
            kSecReturnRef: true
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        if status == errSecSuccess, let item = item {
            return (item as! SecKey)
        }

        var attributes: [CFString: Any] = [
            kSecAttrKeyType: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeySizeInBits: 256,
            kSecAttrTokenID: kSecAttrTokenIDSecureEnclave,
            kSecPrivateKeyAttrs: [
                kSecAttrIsPermanent: true,
                kSecAttrApplicationTag: tag
            ]
        ]
        // Hardcoded access control
        let accessControl = SecAccessControlCreateWithFlags(
            nil,
            kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            .privateKeyUsage,
            nil
        )
        if let ac = accessControl {
            var privateAttrs = attributes[kSecPrivateKeyAttrs] as! [CFString: Any]
            privateAttrs[kSecAttrAccessControl] = ac
            attributes[kSecPrivateKeyAttrs] = privateAttrs
        }

        var error: Unmanaged<CFError>?
        guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
            throw OSSecError(status: errSecParam, message: error?.takeRetainedValue().localizedDescription)
        }
        return privateKey
    }
```
</details>

Commit:
Use hardcoded .privateKeyUsage access control for Secure Enclave
private keys instead of user-provided access control flags. This
prevents "ACL operation is not allowed" errors and eliminates double
authentication prompts.

- Remove accessControl parameter from ensureEnclavePrivateKey
- Add deleteEnclavePrivateKey helper function
- Update deleteAll to delete SE private key when Secure Enclave is enabled
- Remove unused createAccessControl calls

Fixes decryption errors when using Secure Enclave with access control
flags like userPresence or applicationPassword.

Commit:
Format code with melos

Run `melos format` to format all Dart files across the project.
No functional changes.

ahmeddhus commented:
Thanks for the feedback. I've addressed all the issues:
- Fixed SE private key access control: Uses hardcoded .privateKeyUsage instead of user-provided flags, preventing "ACL operation is not allowed" errors.
- Eliminated double authentication: Authentication now only happens once at the data item level.
- Implemented deleteAll cleanup: Added deleteEnclavePrivateKey and integrated it into deleteAll.

This should resolve the decryption errors and double authentication prompts.

@westracer

Commit:
fix: formating

juliansteenbakker commented on flutter_secure_storage_darwin/darwin/flutter_secure_storage_darwin/Sources/flutter_secure_storage_darwin/FlutterSecureStorageDarwinPlugin.swift:
```diff
@@ -165,9 +167,17 @@ public class FlutterSecureStorageDarwinPlugin: NSObject, FlutterPlugin, FlutterS
         )
 
+        // Reuse a single authentication context to avoid multiple prompts per call.
+        if #available(iOS 9.0, macOS 10.12, *) {
```
I think its best to wrap this whole block with a check if useSecureEnclave is true, because otherwise it would change behaviour for the current implementation.

juliansteenbakker commented on flutter_secure_storage_darwin/darwin/flutter_secure_storage_darwin/Sources/flutter_secure_storage_darwin/FlutterSecureStorageDarwinPlugin.swift:
```diff
@@ -5,9 +5,11 @@
 #if os(iOS)
 import Flutter
 import UIKit
+import LocalAuthentication
```
We can move this double declaration above or below the if else statement, since we need it for both iOS and macOS.

juliansteenbakker commented on flutter_secure_storage_darwin/darwin/flutter_secure_storage_darwin/Sources/flutter_secure_storage_darwin/FlutterSecureStorageDarwinPlugin.swift:
```diff
@@ -5,9 +5,11 @@
 #else
 import AppKit
 import FlutterMacOS
+import LocalAuthentication
```
If the above is moved, this one can be removed.

juliansteenbakker reviewed (changes requested):
Thank you again for your hard work! I have two more remarks. If those are solved, i will merge and release it!

Commit:
refactor: Scope LAContext reuse to Secure Enclave and clean up imports

ahmeddhus commented:
Thanks for the review and the detailed suggestions!
 - I’ve updated parseCall so that the LAContext reuse is only enabled when `useSecureEnclave == true`, keeping the behavior unchanged for existing non–Secure Enclave usage.
 - I’ve also moved the LocalAuthentication import above the `#if os(iOS)` / `#else` block and removed the duplicate import, so it’s shared between iOS and macOS.

 Let me know if you’d like me to tweak anything else around the auth context or imports.
 
 @juliansteenbakker

keithhie commented:
Dear team, thank you for the great work in bringing Secure Enclave to be part of this wonderful package. May I know when can we expect this to be released as an official version of the package?

connor-goTenna commented:
Any news on this coming to an official version of?

florianbeaupoilvnv commented:
Hi Team, I would like to thank you for the amazing work on this repository !

This feature is very interesting, and like others here I was wondering when it might be available in an official release. Do you happen to have a roadmap you could share?
